core/data/data.go

```diff
@@ -32,6 +37,66 @@ func SetupPersistence() {
 		}
 	}
 
-	sqliteDatabase, _ := sql.Open("sqlite3", file)
-	_db = sqliteDatabase
+	db, err := sql.Open("sqlite3", file)
+	if err != nil {
+		return err
+	}
+
+	if _, err := db.Exec(`CREATE TABLE IF NOT EXISTS config (
+		"key" string NOT NULL PRIMARY KEY,
+		"value" TEXT
+	);`); err != nil {
+		return err
+	}
+
+	var version int
+	err = db.QueryRow("SELECT value FROM config WHERE key='version'").
+		Scan(&version)
+	if err != nil {
+		if err != sql.ErrNoRows {
+			return err
+		}
+
+		// fresh database: initialize it with the current schema version
+		_, err := db.Exec("INSERT INTO config(key, value) VALUES(?, ?)", "version", schemaVersion)
+		if err != nil {
+			return err
+		}
+		version = schemaVersion
+	}
+
+	// is database from a newer Owncast version?
+	if version > schemaVersion {
+		return fmt.Errorf("incompatible database version %d (versions up to %d are supported)",
+			version, schemaVersion)
+	}
+
+	// is database schema outdated?
+	if version < schemaVersion {
+		if err := migrateDatabase(db, version, schemaVersion); err != nil {
+			return err
+		}
+	}
+
+	_db = db
+	return nil
+}
+
+func migrateDatabase(db *sql.DB, from, to int) error {
+	fmt.Printf("Migrating database from version %d to %d\n", from, to)
+	for v := from; v < to; v++ {
+		switch v {
+		case 0:
```

Comment:
I wonder for the future, if instead of continuing to grow this switch statement for every migration, if we could create a directory of migrations and run them, by name, dynamically.  Something like a `runMigration0(), runMigration1()`, etc.  I know Go isn't a big fan of dynamic programming, but could be a nice way to organize this code.

Not that we need this now, but we could always think about adding it later if we think might be a worthwhile approach.

Reply:
That's one way, even though the code isn't too nice with the use of `reflect` then. Another alternative I've used before is a slice of `[]string`s, containing the SQL commands for each migration. You can then simply range over that `[][]string` and access each version's migration steps by index:

```go
for _, v := range dbMigrations[from:to] {
    for _, s := range v {
        db.Exec(s)
        ...
    }
}
```

Reply:
Oh mixing the two ideas makes a good solution:
A slice of references to functions, so the migration logic still doesn't have to live inline.

```
for _, f := range v {
  f()
```

Just ideas, we can come back to that some other time haha.